File: LeetCode75/array_and_string/reverse_vowels_of_a_string/main.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        VOWELS = {'A', 'E', 'I', 'O', 'U', 'a', 'e', 'i', 'o', 'u'}
        left = 0
        right = len(s) - 1
        reversed = list(s)
        while (left < right):
            while reversed[left] not in VOWELS and left < right:
                left += 1
            while reversed[right] not in VOWELS and left < right:
                right -= 1
            if (reversed[left] in VOWELS and reversed[right] in VOWELS):
                reversed[left], reversed[right] = reversed[right], reversed[left]
            left += 1
            right -= 1
        return "".join(reversed)


if __name__ == "__main__":
    sln = Solution()
    result = sln.reverseVowels("leetcode")
    print(result)

File: LeetCode75/array_and_string/reverse_vowels_of_a_string/main.c
#include <stdbool.h>
#include <string.h>

static bool isVowel(char c)
{
    return c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U' || c == 'a'
           || c == 'e' || c == 'i' || c == 'o' || c == 'u';
}

char* reverseVowels(char* s)
{
    size_t left = 0;
    size_t len = strlen(s);
    size_t right = len - 1;
    while (left < right)
    {
        while (left < right && !isVowel(s[left]))
        {
            left++;
        }
        while (left < right && !isVowel(s[right]))
        {
            right--;
        }
        if (isVowel(s[left]) && isVowel(s[right]))
        {
            char tmp = s[left];
            s[left] = s[right];
            s[right] = tmp;
        }
        left++;
        right = (right == 0) ? 0 : right - 1;
    }
    return s;
}

#include <stdio.h>

int main()
{
    char s[] = "aeiou";
    reverseVowels(s);
    printf("%s\n", s);
}